xsnap/sources/xsnap-worker.c

```diff
@@ -254,6 +255,14 @@ static char *renderTimestamps() {
 	return timestampBuffer;
 }
 
+static void fxSigPipeHandler(int sigNum)
+{
+	if (sigNum == SIGPIPE) {
+		fprintf(stderr, "Caught SIGPIPE. Has parent died?\n");
+		c_exit(E_IO_ERROR);
```

Comment:
I guess technically the SIGPIPE would happen in the middle of a write. Wondering if there is a "cleaner" way to handle it than exiting.

Reply:
Note that this is a current behavior: the default SIGPIPE handler causes immediate process exit.
It also matches pipe write error handling in many other places, for example [xsnap-worker.c:630](https://github.com/agoric-labs/xsnap-pub/blob/a9738b2799a614e6fda5698c8f3cd6e29569ebf8/xsnap/sources/xsnap-worker.c#L630)

Reply:
> It also matches pipe write error handling in many other places, for example [xsnap-worker.c:630](https://github.com/agoric-labs/xsnap-pub/blob/a9738b2799a614e6fda5698c8f3cd6e29569ebf8/xsnap/sources/xsnap-worker.c#L630)

I believe these other places are all when JS is quiescent.


> Note that this is the current behavior: the default SIGPIPE handler causes immediate process exit.

Understood, but since we introduce explicit handling, I was wondering if we could or should do something different than an explicit version of the current behavior. Being a signal handler however it might be difficult to wire it with the place the write happened.

Reply:
> I believe these other places are all when JS is quiescent.

Quiescent JS cannot initiate a write to a pipe by definition. In my view, the "JS transaction" starts with the first byte read by worker from "fromParent" pipe and ends with the last response byte read out by the parent from "toParent" pipe. As defined, the protocol does not allow for any sort of  retry/recovery mechanism, so any interruption in communications in the middle of transaction means that the whole transaction has to be aborted (worker killed) and brought back into a last known good state (restart, reload, and replay).

Also, what can JS do to rectify an issue with a command and control pipe it has no knowledge of? The most I can see is that it could issue a transaction abort to its local database (if it were to actually use one) before shutting it down gracefully. That would save a transaction log replay on restart, but not affect correctness at all.

> Understood, but since we introduce explicit handling, I was wondering if we could or should do something different than an explicit version of the current behavior. Being a signal handler however it might be difficult to wire it with the place the write happened.

I do not fundamentally oppose some other nice[er] options here. However, I do have a strong preference to keep **this** change simple and consistent across the board. Would it make sense to push a complete review of pipe IO error handling strategy and corresponding refactoring into a its own separate issue?

Reply:
May be we should introduce two more pre-registered command handlers: shutdown() and panic()?
This way c-side of xs can more clearly communicate to JS what is happening outside of JS scope. Kind of like SIGTERM/SIGHUP and SIGBUS/SIGSEGV in Unix world.

Reply:
> Quiescent JS cannot initiate a write to a pipe by definition. In my view, the "JS transaction" starts with the first byte read by worker from "fromParent" pipe and ends with the last response byte read out by the parent from "toParent" pipe.

Ah yeah from that point of view, the worker is always active while it tries to write to the pipe. I was taking the definition of JS quiescent as having anything on the JS stack, or pending in the queue. All the other calls to `c_exit` are when writing the result in the big "process command loop" whereas a write also happens within the context of `xs_issueCommand`. In that case, I figured it might be better to `fxAbort` instead of directly exiting.

> what can JS do to rectify an issue with a command and control pipe it has no knowledge of? The most I can see is that it could issue a transaction abort to its local database (if it were to actually use one) before shutting it down gracefully.

Yeah that's kinda the case I had in mind, if in the future we have some teardown we'd prefer doing.

> I do have a strong preference to keep **this** change simple and consistent across the board. Would it make sense to push a complete review of pipe IO error handling strategy and corresponding refactoring into a its own separate issue?

Agreed, in which case `c_exit` for EOF during pipe read would be consistent. Let's file an issue to reconsider and use a "clean" teardown when the JS VM is not quiescent.